src/Common/FailPoint.cpp

```diff
@@ -0,0 +1,191 @@
+#include <Common/Exception.h>
+#include <Common/FailPoint.h>
+
+#include <boost/core/noncopyable.hpp>
+#include <condition_variable>
+#include <mutex>
+#include <optional>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int LOGICAL_ERROR;
+};
+
+#if FIU_ENABLE
+static struct InitFiu
+{
+    InitFiu()
+    {
+        fiu_init(0);
+    }
+} init_fiu;
+#endif
+
+#define APPLY_FOR_FAILPOINTS_ONCE(M) \
+    M(rmt_commit_zk_fail_after_op)
+
+#define APPLY_FOR_FAILPOINTS(M) \
+    M(dummy_failpoint)
+
+#define APPLY_FOR_PAUSEABLE_FAILPOINTS_ONCE(M) \
+    M(dummy_pausable_failpoint_once)
+
+#define APPLY_FOR_PAUSEABLE_FAILPOINTS(M) \
+    M(dummy_pausable_failpoint)
+
+namespace FailPoints
+{
+#define M(NAME) extern const char(NAME)[] = #NAME "";
+APPLY_FOR_FAILPOINTS_ONCE(M)
+APPLY_FOR_FAILPOINTS(M)
+APPLY_FOR_PAUSEABLE_FAILPOINTS_ONCE(M)
+APPLY_FOR_PAUSEABLE_FAILPOINTS(M)
+#undef M
+}
+
+std::unordered_map<String, std::any> FailPointHelper::fail_point_val;
+std::unordered_map<String, std::shared_ptr<FailPointChannel>> FailPointHelper::fail_point_wait_channels;
+class FailPointChannel : private boost::noncopyable
+{
+public:
+    // wake up all waiting threads when destroy
+    ~FailPointChannel() { notifyAll(); }
```

Comment:
Destructor looks not necessary. You cannot call destructor if wait() is called on the same instance in the other thread.